### PR TITLE
BUG: fix missing sort keyword for PeriodIndex.join

### LINKED ISSUE #16541: BUG: Joining a DataFrame with a PeriodIndex fails
#### Code Sample
```python
In [19]: dates = pd.period_range('20100101','20100105', freq='D')

In [20]: weights = pd.DataFrame(np.random.randn(5, 5), index=dates, columns = ['g1_%d' % x for x in range(5)])

In [21]: weights.join(pd.DataFrame(np.random.randn(5,5), index=dates, columns = ['g2_%d' % x for x in range(5)]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-2fdb8b02f5a4> in <module>()
      1 weights.join(
----> 2             pd.DataFrame(np.random.randn(5,5), index=dates, columns = ['g2_%d' % x for x in range(5)]))

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in join(self, other, on, how, lsuffix, rsuffix, sort)
   4765         # For SparseDataFrame's benefit
   4766         return self._join_compat(other, on=on, how=how, lsuffix=lsuffix,
-> 4767                                  rsuffix=rsuffix, sort=sort)
   4768
   4769     def _join_compat(self, other, on=None, how='left', lsuffix='', rsuffix='',

/usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _join_compat(self, other, on, how, lsuffix, rsuffix, sort)
   4780             return merge(self, other, left_on=on, how=how,
   4781                          left_index=on is None, right_index=True,
-> 4782                          suffixes=(lsuffix, rsuffix), sort=sort)
   4783         else:
   4784             if on is not None:

/usr/local/lib/python2.7/dist-packages/pandas/core/reshape/merge.pyc in merge(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator)
     52                          right_index=right_index, sort=sort, suffixes=suffixes,
     53                          copy=copy, indicator=indicator)
---> 54     return op.get_result()
     55
     56

/usr/local/lib/python2.7/dist-packages/pandas/core/reshape/merge.pyc in get_result(self)
    567                 self.left, self.right)
    568
--> 569         join_index, left_indexer, right_indexer = self._get_join_info()
    570
    571         ldata, rdata = self.left._data, self.right._data

/usr/local/lib/python2.7/dist-packages/pandas/core/reshape/merge.pyc in _get_join_info(self)
    720             join_index, left_indexer, right_indexer = \
    721                 left_ax.join(right_ax, how=self.how, return_indexers=True,
--> 722                              sort=self.sort)
    723         elif self.right_index and self.how == 'left':
    724             join_index, left_indexer, right_indexer = \

TypeError: join() got an unexpected keyword argument 'sort'

```

It seems the sort kwarg is invalid, but the internals are passing it in regardless

#### Output of ``pd.show_versions()``

<details>
In [22]: pd.show_versions()
/usr/local/lib/python2.7/dist-packages/xarray/core/formatting.py:16: FutureWarning: The pandas.tslib module is deprecated and will be removed in a future version.
  from pandas.tslib import OutOfBoundsDatetime

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.13-moby
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.7
pip: 9.0.1
setuptools: 35.0.2
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: 0.9.2
IPython: 5.3.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.0
tables: None
numexpr: 2.6.2
feather: None
matplotlib: 2.0.1
openpyxl: None
xlrd: None
xlwt: 1.2.0
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: 0.1.6
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by erbian**
- [x] closes #16541
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can you a whatsnew note (0.21.0), reshaping bug fixes

**Comment by erbian** [comment]
added the note.  looks like the circleci failed on a conda HTTP timeout.  can someone kick off again (i don't have privileges).

**Comment by MaximilianR** [comment]
All good here? @jreback

**Comment by jorisvandenbossche** [comment]
This is a regression, so should target 0.20.3

@erbian Can you move the whatsnew notice to the 0.20.3 file ?

File: pandas/tests/indexes/period/test_period.py [hunk 1]
```diff
@@ -773,3 +773,9 @@ def test_map(self):
         result = index.map(lambda x: x.ordinal)
         exp = Index([x.ordinal for x in index])
         tm.assert_index_equal(result, exp)
+
+    @pytest.mark.parametrize('how', ['outer', 'inner', 'left', 'right'])
+    def test_join_self(self, how):
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/period/test_period.py]: Can you add this to tests/indexes/common.py using `self.create_index` ? The this should automatically be run for all index types

**Review by jorisvandenbossche — COMMENTED** [review]
It seems at least `TimedeltaIndex` has the same problem.

Could you add the fix for that as well?

File: pandas/tests/indexes/period/test_period.py (hunk 1 above)
Comment (erbian) [inline pandas/tests/indexes/period/test_period.py]: added. Non-unique `CategoricalIndex` was breaking in a non-obvious way in `_join_non_unique` so have a look a that fix in particular.

**Comment by jreback** [comment]
pls ``git  rebase -i upstream/master`` then ``git push yourbrachname yourfork -f``

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -3317,7 +3317,7 @@ def _join_monotonic(self, other, how='left', return_indexers=False):
 
     def _wrap_joined_index(self, joined, other):
         name = self.name if self.name == other.name else None
-        return Index(joined, name=name)
+        return self.__class__(joined, name=name)
```
Comment (jorisvandenbossche) [inline pandas/core/indexes/base.py]: Is there a specific reason you changed this (and the line above that)? (meaning, is there currently something failing due to this?)

**Comment by erbian** [comment]
breaking on bug in `numpy 1.13.0 + python 3.6` where `np.array.sort` no longer handles int + str arrays (Issue here: https://github.com/numpy/numpy/issues/9236).  That method is called in `np.unique`

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (erbian) [inline pandas/core/indexes/base.py]: Non-unique `CategoricalIndex` was failing tests and returning a type `Index` not `CategoricalIndex`

File: pandas/tests/indexes/common.py [hunk 3]
```diff
@@ -937,3 +937,14 @@ def test_empty(self):
         index = self.create_index()
         assert not index.empty
         assert index[:0].empty
+
+    @pytest.mark.parametrize('how', ['outer', 'inner', 'left', 'right'])
+    def test_join_self(self, how):
+        index = self.create_index()
+        joined = index.join(index, how=how)
+
+        if index is index.unique():
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/common.py]: Do you mean to do `if index.is_unique:` ?

Comment (jorisvandenbossche) [inline pandas/tests/indexes/common.py]: Just to check why this is needed: not all indexes created by `self.create_index()` are unique ?

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (erbian) [inline pandas/core/indexes/base.py]: Here is the issue replicated:
```
>>> import pandas as pd
>>> import numpy as np
>>> ix = pd.CategoricalIndex(['a', 'b', 'b', 'c'])
>>> ix
CategoricalIndex([u'a', u'b', u'b', u'c'], categories=[u'a', u'b', u'c'], ordered=False, dtype='category')
>>> mask = ix == -1
>>> mask
array([False, False, False, False], dtype=bool)
>>> np.putmask(ix, mask, ix)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: putmask() argument 1 must be numpy.ndarray, not CategoricalIndex
```

Comment (jorisvandenbossche) [inline pandas/core/indexes/base.py]: Yes, good catch. But can you do this in a separate PR? As this will need some more fixes and more tests (eg for the case where both are not exactly the same)

In this PR you can just for now skip the test in case the index is a categorical index

File: pandas/tests/indexes/common.py (hunk 3 above)
Comment (erbian) [inline pandas/tests/indexes/common.py]: yes some are not unique and the result is some modified cartesian product of the two so the test was failing for that case.  So i modified the test to check that at least an index of the same type was returned (potentially a more rigorous test should be done maybe in`tests/indexes/period.py`?).

Comment (erbian) [inline pandas/tests/indexes/common.py]: Using `index.is_unique` will cause this test to fail in the  event that an index `ix` is unique but `ix is ix` returns `False`.  That may be a bug (it's happening with `TimedeltaIndex` currently)?

Comment (jorisvandenbossche) [inline pandas/tests/indexes/common.py]: Do you have a reproducible example with TimedeltaIndex that fails? I can't reproduce it

Comment (erbian) [inline pandas/tests/indexes/common.py]: it's only occuring for right joins:

```
(Pdb) index.join(index) is index
True
(Pdb) index.join(index, how='right') is index
False
(Pdb) index
TimedeltaIndex(['0 days 01:00:00', '1 days 01:00:00', '2 days 01:00:00',
                '3 days 01:00:00', '4 days 01:00:00'],
               dtype='timedelta64[ns]', freq='D')

(Pdb) index.copy() is index
False
(Pdb) index = pd.Index([1,2,3])
(Pdb) index.join(index, how='right')
Int64Index([1, 2, 3], dtype='int64')
(Pdb) index.join(index, how='right') is index
True
(Pdb) index is index.copy()
False

```

Are we are making a copy somewhere in the right-join path?

Comment (jorisvandenbossche) [inline pandas/tests/indexes/common.py]: Ah, that I can reproduce. It is a bit strange this only happens for TimedeltaIndex (might be indication of a bug)

You can also test that `index` and ``joined`` are equal instead of identical.

**Comment by jorisvandenbossche** [comment]
@erbian Do you have time to update this? (would be nice to get this into 0.20.3, due for next week)

**Comment by erbian** [comment]
Fairly busy with other things. I can try to get to it over the weekend.

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (erbian) [inline pandas/core/indexes/base.py]: ok.  thought `_wrap_union_result` uses the `self.__class__` approach so I think this should be ok.

**Comment by erbian** [comment]
@jorisvandenbossche are we good here?

File: doc/source/whatsnew/v0.20.3.txt [hunk 4]
```diff
@@ -79,7 +79,7 @@ Sparse
 
 Reshaping
 ^^^^^^^^^
-
+- Fixed missing sort for index ``join`` method (:issue: `16541`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.3.txt]: ``PeriodIndex / TimedeltaIndex.join`` was missing the ``sort=`` kwarg (you can add a ref to the original issue where it was added as well)

File: doc/source/whatsnew/v0.21.0.txt [hunk 5]
```diff
@@ -132,7 +132,6 @@ Reshaping
 ^^^^^^^^^
 
 
-
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: pls revert this file

File: pandas/tests/indexes/common.py [hunk 6]
```diff
@@ -933,3 +933,10 @@ def test_empty(self):
         index = self.create_index()
         assert not index.empty
         assert index[:0].empty
+
+    @pytest.mark.parametrize('how', ['outer', 'inner', 'left', 'right'])
```
Comment (jreback) [inline pandas/tests/indexes/common.py]: can you also add the tests from the original issue (it was on PI) as a specific test.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@erbian can you edit for comments. releasing 0.20.3 tomorrow and want to push this in.

File: doc/source/whatsnew/v0.20.3.txt (hunk 4 above)
Comment (erbian) [inline doc/source/whatsnew/v0.20.3.txt]: ok.  the original issue was `16541` which is already in the note?

**Comment by erbian** [comment]
ok @jreback @jorisvandenbossche I edited for latest comments and cleaned up the PR (4 commits now).  It is rebased on latest upstream/master as well.

**Review by TomAugspurger — APPROVED** [review]

File: doc/source/whatsnew/v0.20.3.txt [hunk 7]
```diff
@@ -82,7 +82,7 @@ Sparse
 
 Reshaping
 ^^^^^^^^^
-
+- PeriodIndex / TimedeltaIndex.join was missing the sort= kwarg (:issue: `16541`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.3.txt]: can you use double-backticks around PeriodIndex & TimedeltaIndex and ``sort=``

File: pandas/tests/indexes/common.py (hunk 6 above)
Comment (jreback) [inline pandas/tests/indexes/common.py]: I mean this test here: https://github.com/pandas-dev/pandas/issues/16541#issue-232356015

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/common.py (hunk 6 above)
Comment (erbian) [inline pandas/tests/indexes/common.py]: do mean mean add the code that generated the issue as a test (joining two randomly generated 5x5 DataFrames)?  I don't really understand what you're asking me to do.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@TomAugspurger I think this is the last PR for 0.20.3. waiting on pass to merge.

**Comment by TomAugspurger** [comment]
Cool. Looks like it's next in line for Appveyor. I'll merge it on green and start the release stuff.

**Comment by jreback** [comment]
thanks @erbian
